rmw_dds_common/include/rmw_dds_common/graph_cache.hpp

```diff
@@ -0,0 +1,501 @@
+    const rmw_gid_t & participant_gid,
+    const std::string & node_name,
+    const std::string & node_namespace);
+
+  /**
+   * Dissociate a writer from a node, and get an update message to be sent.
+   *
+   * \param writer_gid GUID of the data writer.
+   * \param participant_gid GUID of the `Participant`.
+   * \param node_name Name of the node to be added.
+   * \param node_namespace Node namespace.
+   * \return Message to be sent.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_dds_common::msg::ParticipantEntitiesInfo
+  dissociate_writer(
+    const rmw_gid_t & writer_gid,
+    const rmw_gid_t & participant_gid,
+    const std::string & node_name,
+    const std::string & node_namespace);
+
+  /**
+   * Associate a reader with a node, and get an update message to be sent.
+   *
+   * \param reader_gid GUID of the data reader.
+   * \param participant_gid GUID of the `Participant`.
+   * \param node_name Name of the node to be added.
+   * \param node_namespace Node namespace.
+   * \return Message to be sent.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_dds_common::msg::ParticipantEntitiesInfo
+  associate_reader(
+    const rmw_gid_t & reader_gid,
+    const rmw_gid_t & participant_gid,
+    const std::string & node_name,
+    const std::string & node_namespace);
+
+  /**
+   * Dissociate a reader from a node, and get an update message to be sent.
+   *
+   * \param reader_gid GUID of the data reader.
+   * \param participant_gid GUID of the `Participant`.
+   * \param node_name Name of the node to be added.
+   * \param node_namespace Node namespace.
+   * \return Message to be sent.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_dds_common::msg::ParticipantEntitiesInfo
+  dissociate_reader(
+    const rmw_gid_t & reader_gid,
+    const rmw_gid_t & participant_gid,
+    const std::string & node_name,
+    const std::string & node_namespace);
+
+  /**
+   * @}
+   * \defgroup introspection_api
+   * Methods used to introspect the GraphCache.
+   * @{
+   */
+
+  /**
+   * Get the number of publishers of a topic.
+   *
+   * \param[in] topic_name Name of the topic.
+   * \param[out] count The result will be populated there.
+   *
+   * \return RMW_RET_INVALID_ARGUMENT if count is `nullptr`, or
+   * \return RMW_RET_ERROR if an unexpected error take place, or
+   * \return RMW_RET_OK.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  get_writer_count(
+    const std::string & topic_name,
+    size_t * count) const;
+
+  /**
+   * Get the number of subscriptions of a topic.
+   *
+   * \param[in] topic_name Name of the topic.
+   * \param[out] count The result will be populated there.
+   *
+   * \return RMW_RET_INVALID_ARGUMENT if count is `nullptr`, or
+   * \return RMW_RET_ERROR if an unexpected error take place, or
+   * \return RMW_RET_OK.
+   */
+  RMW_DDS_COMMON_PUBLIC
+  rmw_ret_t
+  get_reader_count(
+    const std::string & topic_name,
+    size_t * count) const;
+
+  using DemangleFunctionT = std::string (*)(const std::string &);
```

Comment:
I think that `std::function<>` is not a zero cost abstraction over a function pointer, and should be used only as a container. Although, I understand the overhead is minumum and my opinion is a bit exagerated). Another option is to add a template parameter.

Reply:
> I think that `std::function<>` is not a zero cost abstraction over a function pointer

Is that true? I thought it was something that the compiler could optimize, and in the worst case, it was not a zero-cost abstraction but the overhead was very small.

On the flip side, it allows you to use a `std::bind` or lambda with the functions that use this typedef as an argument type, in the future if desired, so it seems like something nice to have.

My take would be, if you want to avoid the abstraction overhead, use a template argument, but otherwise use `std::function<>`. Using a C-style function pointer to avoid the overhead isn't really idiomatically C++.

Reply:
> On the flip side, it allows you to use a std::bind or lambda with the functions that use this typedef as an argument type, in the future if desired, so it seems like something nice to have.

That's true.

> My take would be, if you want to avoid the abstraction overhead, use a template argument, but otherwise use std::function<>. Using a C-style function pointer to avoid the overhead isn't really idiomatically C++.

Fair enough.

> Is that true? I thought it was something that the compiler could optimize, and in the worst case, it was not a zero-cost abstraction but the overhead was very small.

I don't think that the compiler can optimize out type erasure in a function call.